Feat!: Dev-only VDE mode (#5087)
TobikoData/sqlmesh: docs/guides/configuration.md

This has a downside that now it's much more difficult to determine which table corresponds to which model by just looking at the database with a SQL client. However, the table names have a predictable length so there are no longer any surprises with identfiers exceeding the max length at the physical layer.

#### Virtual Data Environment modes

By default, Virtual Data Environments (VDE) are applied across both development and production environments. This allows SQLMesh to reuse physical tables when appropriate, even when promoting from development to production.

However, users may sometimes prefer their production environment to be non-virtual. The non-exhaustive list of reasons may include:

- Integration with third-party tools and platforms, such as data catalogs, may not work well with the virtual view layer that SQLMesh imposes by default

> - A desire to rely on time travel features provided by cloud data warehouses such as BigQuery, Snowflake, and Databricks

Reviewer: ```suggestion
- A desire to rely on time travel features provided by cloud data warehouses such as BigQuery, Snowflake, and Databricks. This is mainly for hot fixes in production.
```

To mitigate this, SQLMesh offers an alternative 'dev-only' mode for using VDE. It can be enabled in the project configuration like so:

=== "YAML"

    ```yaml linenums="1"
    virtual_environment_mode: dev_only
    ```

=== "Python"

    ```python linenums="1"
    from sqlmesh.core.config import Config

    config = Config(
        virtual_environment_mode="dev_only",
    )
    ```

> As the name suggests, 'dev-only' mode means that VDE is applied only in development environments, while in production, model tables and views are updated directly, bypassing the virtual layer. This also means that physical tables in production will be created using the original, unversioned model names. Users will still benefit from VDE and data reuse across development environments.

Reviewer: ```suggestion
'dev-only' mode means that VDE is applied only in development environments. While in production, model tables and views are updated directly and bypass the virtual layer. This also means that physical tables in production will be created using the original, **unversioned** model names. Users will still benefit from VDE and data reuse across development environments.
```

Please note the following tradeoffs when enabling this mode:

- All data inserted in development environments is used only for [preview](../concepts/plans.md#data-preview-for-forward-only-changes) and will **not** be reused in production
- Reverting a model to a previous version will be applied going forward and may require an explicit data restatement

!!! warning

>     Switching the mode for an existing project will result in a complete rebuild of all models in the project. Refer to the [Table Migration Guide](./table_migration.md) to migrate existing tables without rebuilding them from scratch.

Reviewer: ```suggestion
    Switching the mode for an existing project will result in a **complete rebuild** of all models in the project. Refer to the [Table Migration Guide](./table_migration.md) to migrate existing tables without rebuilding them from scratch.
```
